File: vision_fullcam/rules/worker_count.py
from typing import List
from vision_fullcam.rules.base import Rule, RuleContext, Event, Debounce
from vision_fullcam.config import Config


class InsufficientWorkerCountRule(Rule):
    name = "insufficient_worker_count"

    def __init__(self, cfg: Config):
        self._db = Debounce(cfg.worker_missing_sec, cfg.cooldown_sec)

    def is_active(self, ctx: RuleContext) -> bool:
        return ctx.state.person_on_ladder()

    def evaluate(self, ctx: RuleContext) -> List[Event]:
        now  = ctx.timestamp
        site = ctx.state.site
        cond = site.any_ladder and (site.person_count < 2)

        if self._db.check(now, cond):
            return [Event(
                self.name, "high", None, now,
                {"person_count": site.person_count}
            )]

        return []

File: frontend/mobile/src/vision/rules/workerCount.ts
import type { SafetyEvent } from '../types';
import type { VisionConfig } from '../config';
import { Debounce } from './base';
import type { Rule, RuleContext } from './base';

export class InsufficientWorkerCountRule implements Rule {
  name = 'insufficient_worker_count';
  private db: Debounce;

  constructor(cfg: VisionConfig) {
    this.db = new Debounce(cfg.workerMissingSec, cfg.cooldownSec);
  }

  evaluate(ctx: RuleContext): SafetyEvent[] {
    const now = ctx.timestamp;
    const cond =
      ctx.state.site.anyLadder && ctx.state.site.personCount < 2;

    if (this.db.check(now, cond)) {
      return [
        {
          label: this.name,
          severity: 'high',
          targetId: null,
          ts: now,
          info: { person_count: ctx.state.site.personCount },
        },
      ];
    }
    return [];
  }
}
